### PR TITLE
Improve test logs

### PR DESCRIPTION
**Pull request by halter73**
- Include hosting logs in some tests that where previously missing them
- Prevent duplicate logs from Mock CallBase an CompositeKestrelTrace
- Add file logging to HTTP/2 "unit" tests

I'm checking some changes I made locally when doing flaky test investigation.

### CODE REVIEW

File: test/Kestrel.Core.Tests/Http2StreamTests.cs [hunk 1]
```diff
@@ -93,127 +52,6 @@ public Http2StreamTests()
...
-        public void Dispose()
-        {
-            _pair.Application.Input.Complete();
-            _pair.Application.Output.Complete();
-            _pair.Transport.Input.Complete();
-            _pair.Transport.Output.Complete();
-            _memoryPool.Dispose();
-        }
-
-        void IHttpHeadersHandler.OnHeader(Span<byte> name, Span<byte> value)
```
Comment (Tratcher) [inline test/Kestrel.Core.Tests/Http2StreamTests.cs]: Can you also move WaitForConnectionErrorAsync, WaitForStreamErrorAsync, etc.?

**Comment by Tratcher** [comment]
CriticalErrorLoggedIfApplicationDoesntComplete is failing on several builds.

File: test/Kestrel.InMemory.FunctionalTests/Kestrel.InMemory.FunctionalTests.csproj [hunk 2]
```diff
@@ -20,8 +19,8 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNetCore.Http.Abstractions" Version="$(MicrosoftAspNetCoreHttpAbstractionsPackageVersion)" />
-    <PackageReference Include="Microsoft.AspNetCore.Testing" Version="$(MicrosoftAspNetCoreTestingPackageVersion)" />
     <PackageReference Include="Microsoft.Extensions.Logging.Testing" Version="$(MicrosoftExtensionsLoggingTestingPackageVersion)" />
+    <PackageReference Include="Microsoft.AspNetCore.Testing" Version="$(MicrosoftAspNetCoreTestingPackageVersion)" />
```
Comment (JunTaoLuo) [inline test/Kestrel.InMemory.FunctionalTests/Kestrel.InMemory.FunctionalTests.csproj]: ?

File: test/Kestrel.Transport.Libuv.FunctionalTests/Kestrel.Transport.Libuv.FunctionalTests.csproj [hunk 3]
```diff
@@ -26,8 +25,8 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNetCore.Http.Abstractions" Version="$(MicrosoftAspNetCoreHttpAbstractionsPackageVersion)" />
-    <PackageReference Include="Microsoft.AspNetCore.Testing" Version="$(MicrosoftAspNetCoreTestingPackageVersion)" />
     <PackageReference Include="Microsoft.Extensions.Logging.Testing" Version="$(MicrosoftExtensionsLoggingTestingPackageVersion)" />
+    <PackageReference Include="Microsoft.AspNetCore.Testing" Version="$(MicrosoftAspNetCoreTestingPackageVersion)" />
```
Comment (JunTaoLuo) [inline test/Kestrel.Transport.Libuv.FunctionalTests/Kestrel.Transport.Libuv.FunctionalTests.csproj]: why move these references down?

File: test/Kestrel.Transport.Sockets.FunctionalTests/Kestrel.Transport.Sockets.FunctionalTests.csproj [hunk 4]
```diff
@@ -25,8 +24,8 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNetCore.Http.Abstractions" Version="$(MicrosoftAspNetCoreHttpAbstractionsPackageVersion)" />
-    <PackageReference Include="Microsoft.AspNetCore.Testing" Version="$(MicrosoftAspNetCoreTestingPackageVersion)" />
     <PackageReference Include="Microsoft.Extensions.Logging.Testing" Version="$(MicrosoftExtensionsLoggingTestingPackageVersion)" />
+    <PackageReference Include="Microsoft.AspNetCore.Testing" Version="$(MicrosoftAspNetCoreTestingPackageVersion)" />
```
Comment (JunTaoLuo) [inline test/Kestrel.Transport.Sockets.FunctionalTests/Kestrel.Transport.Sockets.FunctionalTests.csproj]: ?

File: test/Kestrel.Core.Tests/TestHelpers/Http2TestBase.cs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
+using Microsoft.AspNetCore.Server.Kestrel.Transport.Abstractions.Internal;
+using Microsoft.AspNetCore.Testing;
+using Microsoft.Net.Http.Headers;
+using Moq;
+using Xunit.Abstractions;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Tests
+{
+    public class Http2TestBase : TestApplicationErrorLoggerLoggedTest, IDisposable, IHttpHeadersHandler
```
Comment (JunTaoLuo) [inline test/Kestrel.Core.Tests/TestHelpers/Http2TestBase.cs]: :+1:

**Comment by JunTaoLuo** [comment]
I believe we want to wait until release/2.2 is open for preview2

**Comment by halter73** [comment]
I'm moving the Http2Connection/StreamTests it the InMemory.FunctionalTests project because each Http2Connection/StreamTest was consistently failing with an error similar to the following only when running `build.cmd`. Running `dotnet test` in the Core.Tests project or `build.cmd -ci` did not reproduce the error. The error also only occurred in the "UnitTests/net461" test group.


```

  Starting test group: UnitTests/netcoreapp2.2
  Microsoft (R) Test Execution Command Line Tool Version 15.7.0
  Copyright (c) Microsoft Corporation.  All rights reserved.

  Starting test execution, please wait...

  Total tests: 6479. Passed: 6479. Failed: 0. Skipped: 0.
  Test Run Successful.
  Test execution time: 22.4227 Seconds

  Starting test group: UnitTests/net461
  Microsoft (R) Test Execution Command Line Tool Version 15.7.0
  Copyright (c) Microsoft Corporation.  All rights reserved.

  Starting test execution, please wait...
  [xUnit.net 00:00:06.91]     HEADERS_Received_Path_DecodedAndNormalized(input: "/", expected: "/") [FAIL]
  [xUnit.net 00:00:06.92]     HEADERS_Received_Path_DecodedAndNormalized(input: "/a%5E", expected: "/a^") [FAIL]
  [xUnit.net 00:00:06.93]     HEADERS_Received_Path_DecodedAndNormalized(input: "/a%b", expected: "/a%b") [FAIL]
  [xUnit.net 00:00:06.93]     HEADERS_Received_Path_DecodedAndNormalized(input: "/a/b/c/../../../../d", expected: "/d") [FAIL]
  Failed   HEADERS_Received_Path_DecodedAndNormalized(input: "/", expected: "/")
  Error Message:
   System.AggregateException : One or more errors occurred.
  ---- System.NullReferenceException : Object reference not set to an instance of an object.
  ---- System.NullReferenceException : Object reference not set to an instance of an object.
  Stack Trace:

  ----- Inner Stack Trace #1 (System.NullReferenceException) -----
     at Microsoft.AspNetCore.Server.Kestrel.Core.Tests.Http2TestBase.<InitializeConnectionAsync>d__34.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
     at Microsoft.AspNetCore.Server.Kestrel.Core.Tests.Http2StreamTests.<HEADERS_Received_Path_DecodedAndNormalized>d__7.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  ----- Inner Stack Trace #2 (System.NullReferenceException) -----
     at Microsoft.Extensions.Logging.Testing.LoggedTestBase.Dispose()
     at Microsoft.AspNetCore.Server.Kestrel.Core.Tests.Http2TestBase.Dispose()
     at ReflectionAbstractionExtensions.DisposeTestClass(ITest test, Object testClass, IMessageBus messageBus, ExecutionTimer timer, CancellationTokenSource cancellationTokenSource)

...

  Total tests: 6478. Passed: 6216. Failed: 262. Skipped: 0.
  Test Run Failed.
  Test execution time: 17.4629 Seconds
C:\Users\shalter\.dotnet\buildtools\korebuild\2.2.0-preview1-20180807.2\modules\vstest\module.targets(108,5): error : Test group UnitTests/net461 failed with exit code '1'. [C:\Users\shalter\.dotnet\buildtools\korebuild\2.2.0-preview1-20180807.2\KoreBuild.proj]

```

@JunTaoLuo and I verified this issue occurred on multiple machines, and the proximate cause was `Http2TestBase.Initialize()` not being called. Moving to the InMemory.FunctionalTests project fixed the issue. We compared the Core.Tests and InMemory.FunctionalTests projects and couldn't determine why this is an issue in Core.Tests and not InMemory.FunctionalTests . I'm merging anyway, since I'm about the leave on vacation and don't want to deal with merge conflicts.

**Pull request merged**